crates/libs/windows/src/core/factory_cache.rs

```diff
@@ -59,50 +59,62 @@ pub fn factory<C: RuntimeName, I: Interface>() -> Result<I> {
             if code == CO_E_NOTINITIALIZED {
                 if let Ok(mta) = delay_load(b"ole32.dll\0", b"CoIncrementMTAUsage\0") {
                     let mta: CoIncrementMTAUsage = core::mem::transmute(mta);
-                    let mut _cookie = core::ptr::null_mut();
-                    let _ = mta(&mut _cookie);
+                    let mut cookie = core::ptr::null_mut();
+                    let _ = mta(&mut cookie);
                 }
 
                 // Now try a second time to get the activation factory via the OS.
                 code = function(core::mem::transmute_copy(&name), &I::IID, &mut factory as *mut _ as *mut _);
             }
 
             code
-        } else {
-            CLASS_E_CLASSNOTAVAILABLE
-        };
-
-        // If this succeeded then return the resulting factory interface.
-        if code.is_ok() {
-            return code.and_some(factory);
         }
+    } else {
+        CLASS_E_CLASSNOTAVAILABLE
+    };
 
-        // If not, first capture the error information from the failure above so that we
-        // can ultimately return this error information if all else fails.
-        let original: Error = code.into();
-
-        // Now attempt to find the factory's implementation heuristically.
-        let mut path = C::NAME;
+    // If this succeeded then return the resulting factory interface.
+    if code.is_ok() {
+        return code.and_some(factory);
+    }
 
-        // Remove the suffix until a match is found. For example, if the class name is
-        // "A.B.TypeName" then the attempted load order will be:
-        //   1. A.B.TypeName.dll
-        //   2. A.B.dll
-        //   3. A.dll
-        while let Some(pos) = path.rfind('.') {
-            path = &path[..pos];
+    // If not, first capture the error information from the failure above so that we
+    // can ultimately return this error information if all else fails.
+    let original: Error = code.into();
 
-            let library = core::slice::from_raw_parts_mut(heap_alloc(path.len() + 5)? as *mut u8, path.len() + 5);
-            library[..path.len()].copy_from_slice(path.as_bytes());
-            library[path.len()..].copy_from_slice(b".dll\0");
+    // Now attempt to find the factory's implementation heuristically.
+    let path = C::NAME;
+    if let Some(i) = search_path(path, |library| unsafe { get_activation_factory(library, &name) }) {
+        i.cast()
+    } else {
+        Err(original)
+    }
+}
 
-            if let Ok(factory) = get_activation_factory(library, &name) {
-                return factory.cast();
-            }
+// Remove the suffix until a match is found appending `.dll\0` at the end
+///
+/// For example, if the class name is
+/// "A.B.TypeName" then the attempted load order will be:
+///   1. A.B.dll
+///   2. A.dll
```

Comment:
@riverar thanks to the test, I found that the implementation of the search heuristic differed from the documentation. The documentation here now matches the behavior. Was the previous documentation correct? Should this first search `A.B.TypeName.dll`?

Reply:
I knew this sounded familiar, hah! I did correct this comment when I was in here https://github.com/microsoft/windows-rs/pull/1658/files#diff-b5b69fd449b1b061e71948bc6db077ea2d34ce8cca5a11aea3de2e12588a2f3eR96-R100

Reply:
I believe this is an artifact from my previous explicit activation support. The comment is OK as-is now.

Here's the supporting documentation https://docs.microsoft.com/en-us/windows/uwp/winrt-components/create-a-windows-runtime-component-in-cppwinrt#naming-best-practice-for-windows-runtime-component-dlls

> [...] The logic is to remove the class name (from that fully-qualified class name), and then look for a DLL named for the full namespace that remains. If that's not found, remove the most-specific segment name, and repeat.